Wrappers/Python/cil/optimisation/utilities/preconditioner.py

```diff
@@ -0,0 +1,199 @@
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+# Authors:
+# - CIL Developers, listed at: https://github.com/TomographicImaging/CIL/blob/master/NOTICE.txt
+
+
+from abc import ABC, abstractmethod
+import numpy as np
+
+
+class Preconditioner(ABC):
+
+    r"""
+    Abstract base class for Preconditioner objects. The `apply` method of this class takes an initialised CIL function as an argument and modifies a provided  `gradient`.
+   
+    
+    Methods
+    -------
+    apply(x)
+        Abstract method to call the preconditioner.
+    """
+
+    def __init__(self):
+        '''Initialises the preconditioner
+        '''
+        pass
+
+    @abstractmethod
+    def apply(self, algorithm, gradient, out):
+        r"""
+        Abstract method to apply the preconditioner.
+
+        Parameters
+        ----------
+        algorithm : Algorithm
+            The algorithm object.
+    
+        gradient : DataContainer
+            The calculated gradient to modify
+        out : DataContainer, 
+            Container to fill with the modified gradient 
+        
+        Returns
+        -------
+        DataContainer
+            The modified gradient
+            
+        """
+        pass
+
+
+class Sensitivity(Preconditioner):
+
+    r"""
+    Sensitivity preconditioner class. 
+
+    In each call to the preconditioner the `gradient` is multiplied by :math:` 1/(A^T \mathbf{1})` where :math:`A` is an operator, :math:`\mathbf{1}` is an object in the range of the operator filled with ones.
+
+    Parameters
+    ----------
+    operator : CIL Operator 
+        The operator used for sensitivity computation.
+
+    """
+
+    def __init__(self, operator):
+
+        super(Sensitivity, self).__init__()
+        self.operator = operator
+
+
+        self.compute_preconditioner_matrix()
+
+    def compute_preconditioner_matrix(self):
+        r"""
+        Compute the sensitivity. :math:`A^T \mathbf{1}` where :math:`A` is the operator and :math:`\mathbf{1}` is an object in the range of the operator filled with ones.        
+        Then perform safe division by the sensitivity to store the preconditioner array :math:` 1/(A^T \mathbf{1})`
+        """
+        self.array = self.operator.adjoint(
+            self.operator.range_geometry().allocate(value=1.0))
+        
+        self.operator.range_geometry().allocate(value=1.0).divide(
+                self.array, where=np.abs(self.array.as_array()) > 0, out=self.array)
```

Comment:
`where` will not be recognised by [SIRF](https://github.com/SyneRBI/SIRF/blob/9690feb5cff4694c2edcbe84cc4cb17e4bb85dd7/src/common/SIRF.py#L374).